sw/device/lib/dif/dif_pinmux.c

```diff
@@ -0,0 +1,147 @@
+
+  return kDifPinmuxInitErrorNone;
+}
+
+dif_pinmux_registers_lock_error_t dif_pinmux_registers_lock(
+    dif_pinmux_t *pinmux) {
+  if (pinmux == NULL) {
+    return kDifPinmuxRegistersLockErrorInvalidArgument;
+  }
+
+  mmio_region_write32(pinmux->base_addr, PINMUX_REGEN_REG_OFFSET, 0);
+  pinmux->write_enabled = false;
+
+  return kDifPinmuxRegistersLockErrorNone;
+}
+
+dif_pinmux_insel_set_error_t dif_pinmux_insel_set(
+    const dif_pinmux_t *pinmux, dif_pinmux_peripheral_in_t peripheral_input,
+    dif_pinmux_insel_t mio_select) {
+  if (pinmux == NULL) {
+    return kDifPinmuxInselSetErrorInvalidArgument;
+  }
+
+  if (!pinmux->write_enabled) {
+    return kDifPinmuxInselSetErrorRegistersLocked;
+  }
+
+  if (mio_select < kDifPinmuxInselTieToZero ||
+      mio_select > kDifPinmuxInselPadLast) {
+    return kDifPinmuxInselSetErrorBadMioSelect;
+  }
+
+  // Register offset based on a given peripheral input and MIO select.
+  uint32_t register_index =
+      peripheral_input / PINMUX_PERIPH_INSEL_IN_FIELDS_PER_REG;
+  ptrdiff_t register_offset = PINMUX_PERIPH_INSEL0_REG_OFFSET +
+                              (register_index * PINMUX_REGISTER_WIDTH_IN_BYTES);
+
+  uint32_t field_index =
+      peripheral_input % PINMUX_PERIPH_INSEL_IN_FIELDS_PER_REG;
+  uint32_t field_offset = field_index * PINMUX_PERIPH_INSEL_IN_FIELD_WIDTH;
+
+  // INSEL field masks are the same for all of the fields.
+  bitfield_field32_t field = {
+      .mask = PINMUX_PERIPH_INSEL0_IN0_MASK,
+      .index = field_offset,
+      .value = mio_select,
+  };
+  mmio_region_nonatomic_set_field32(pinmux->base_addr, register_offset, field);
+
+  return kDifPinmuxInselSetErrorNone;
+}
+
+dif_pinmux_outsel_set_error_t dif_pinmux_outsel_set(
+    const dif_pinmux_t *pinmux, dif_pinmux_mio_in_t mio_input_pad,
+    dif_pinmux_outsel_t peripheral_select) {
+  if (pinmux == NULL) {
+    return kDifPinmuxOutselSetErrorInvalidArgument;
+  }
+
+  if (!pinmux->write_enabled) {
+    return kDifPinmuxOutselSetErrorRegistersLocked;
+  }
+
+  if (mio_input_pad < kDifPinmuxMioInPadFirst ||
+      mio_input_pad > kDifPinmuxMioInPadLast) {
+    return kDifPinmuxOutselSetErrorBadMio;
+  }
+
+  // Register offset based on a given peripheral output and MIO select.
+  uint32_t register_index =
+      mio_input_pad / PINMUX_MIO_OUTSEL_OUT_FIELDS_PER_REG;
+  ptrdiff_t register_offset = PINMUX_MIO_OUTSEL0_REG_OFFSET +
+                              (register_index * PINMUX_REGISTER_WIDTH_IN_BYTES);
+
+  uint32_t field_index = mio_input_pad % PINMUX_MIO_OUTSEL_OUT_FIELDS_PER_REG;
+  uint32_t field_offset = field_index * PINMUX_MIO_OUTSEL_OUT_FIELD_WIDTH;
+
+  // OUTSEL field masks are the same for all of the fields.
+  bitfield_field32_t field = {
+      .mask = PINMUX_MIO_OUTSEL0_OUT0_MASK,
+      .index = field_offset,
+      .value = peripheral_select,
+  };
+  mmio_region_nonatomic_set_field32(pinmux->base_addr, register_offset, field);
+
+  return kDifPinmuxOutselSetErrorNone;
+}
```

Comment:
Shouldn't we have functions for reading the current configuration?

Reply:
Hmmm, I see your point, I guess it makes sense to return these value in bulk (all values of insel, all values of outsel).

Not sure about the API, we might need to expose `PINMUX_PERIPH_INSEL_MULTIREG_COUNT` and `PINMUX_PERIPH_OUTSEL_MULTIREG_COUNT` to allow caller create a buffer to hold all values. This buffer then can be passed into a `*get` function.

It might be worth waiting until we auto generate pinmux defines in `top_earlgrey`, all this should be easier then.